### DISCUSSION TITLE
How do I create nested tables ?

### DISCUSSION THREAD

**Discussion by AllergicMushroom** [saulpw/visidata#2611] [Q&A] [answered]
How do I create nested tables ?

Visidata can of course read them, I can expand and collapse the columns, but I can't find a way to actually create a table inside of a cell of my top-level table.

I can press enter to get the details of the row, then enter again to get the details of the cell, but if I had two lines there and quit with q, nothing is saved. However, if I follow these steps to modify a cell containing a nested table, the changes are saved.

**Comment by midichef** [comment]
What do you mean by nested table?

It would help if you can demonstrate with a short example file that has a nested table.
And when you try to make a nested table, what data file did you start with, and what are all the keypresses you used? In particular, to make the two lines?

**Reply by AllergicMushroom** [reply]
Currently, the only way I have found to manipulate nested tables are via a text editor, which defeats the purpose of using visidata in the first place.

Consider the following jsonl example read by visidata.

```
{"A": [1, 2, 3]}
```

### Problem 1

If I expand with `(`, I can edit the cells, collapse with `)`, and save, everything works as expected.
If I save without collapsing, the file gets converted to the following.

```
{"A[0]: 1, "A[1]": 2, "A[2]": 3}
```

visidata treats this as three separate columns, uncapable of being collapsed with `)`.
I don't know if this is desired or not.

### Problem 2

I cannot find any sequence of keypresses that allow me to add a column.
For example, I want to start from the first example and have the following as output.

```
{"A": [1, 2, 3, 4]}
```

In the case of multiple rows, I wish to add a column only to A's nested table.

Consider the following example.

```
{"A": [1, 2, 3]}
{"B": [1, 4, 9]}
```

I want to add an entry only to the nested table of A, and have the following ouput.

```
{"A": [1, 2, 3, 4]}
{"B": [1, 4, 9]}
```

**Reply by midichef** [reply, marked as answer]
I see. For nested table, you are referring to cells containing sequences like lists or dictionaries, which can be expanded with `expand-col`.

For your problem 1, that saving expanded columns loses the ability to contract them, I don't know if that's desired behavior. Maybe @saulpw can comment?

For problem 2, where you want to add to the list contained in a cell. You can use `z=` (for `setcell-expr`) to set the current cell to any Python expression. Expressions you could use to add a value are:
`[1,2,3,4]`
`A + [4]`
`curcol + [4]` (reading the value of the current column where the cursor is, instead of using the column name `A`)

If you want to simplify this process, you could make a custom command in your .visidatarc to run `append()` or `extend()` for cells that hold lists:
```
Sheet.addCommand('', 'append-cell-expr', 'cursorValue.append(evalExpr(inputExpr("append expr="), row=cursorRow, curcol=cursorCol))', 'append to current cell')
Sheet.addCommand('', 'extend-cell-expr', 'cursorValue.extend(evalExpr(inputExpr("extend expr="), row=cursorRow, curcol=cursorCol))', 'extend the current cell')
```

Overall, I'd say Visidata's ability to work with nested data structures is quite rough around the edges. If there are improvements you'd like to suggest, I for one would love to hear them. Opening an issue with a feature request would be a good way to start a discussion about specifics.

**Reply by AllergicMushroom** [reply]
I have marked your previous message as the answer, as the `z=` was I what as I was searching for.

As for the nested data structures, I believe the most important thing is to access the underlying data structure when pressing `enter`.

However, if the nested data structure is a table with colums and rows, instead of a simple list, I am unsure how to display it in the view of the 'main' table.
